src/System.Management.Automation/engine/LanguagePrimitives.cs

```diff
@@ -2900,36 +2932,55 @@ private static bool TryScanNumber(string strToConvert, Type resultType, out obje
             }
         }
 
-        private static object ConvertStringToInteger(object valueToConvert,
-                                                     Type resultType,
-                                                     bool recursion,
-                                                     PSObject originalValueToConvert,
-                                                     IFormatProvider formatProvider,
-                                                     TypeTable backupTable)
+        private static object ConvertStringToInteger(
+            object valueToConvert,
+            Type resultType,
+            bool recursion,
+            PSObject originalValueToConvert,
+            IFormatProvider formatProvider,
+            TypeTable backupTable)
         {
             var strToConvert = valueToConvert as string;
             Diagnostics.Assert(strToConvert != null, "Value to convert must be a string");
-            Diagnostics.Assert(IsNumeric(GetTypeCode(resultType)), "Result type must be numeric");
+            Diagnostics.Assert(
+                IsNumeric(GetTypeCode(resultType)) || resultType == typeof(BigInteger),
+                "Result type must be numeric");
 
             if (strToConvert.Length == 0)
             {
                 typeConversion.WriteLine("Returning numeric zero.");
+
+                // BigInteger is not IConvertible and will throw from ChangeType.
+                if (resultType == typeof(BigInteger))
+                {
+                    return BigInteger.Zero;
+                }
+
                 // This is not wrapped in a try/catch because it can't fail.
                 return System.Convert.ChangeType(0, resultType, CultureInfo.InvariantCulture);
             }
 
             typeConversion.WriteLine("Converting to integer.");
-            TypeConverter integerConverter = LanguagePrimitives.GetIntegerSystemConverter(resultType);
+
+            // This will contain null if resultType == typeof(BigInteger)
+            TypeConverter integerConverter = GetIntegerSystemConverter(resultType);
             try
             {
                 if (TryScanNumber(strToConvert, resultType, out object result))
                 {
                     return result;
                 }
-                else
+
+                if (resultType == typeof(BigInteger))
```

Comment:
Why do we need `integerConverter` if we can return after TryScanNumber?
And why do we check `resultType == typeof(BigInteger)` if `integerConverter` could be null already?
I think we should refactor the code.

Reply:
I couldn't see a good way to refactor that without completely diverging the code path, but I'll take another look.

Reply:
Looks like we don't even need the variable, you're right. I'll redo this a bit more.